File: backend/size_range.py
"""Expand clothing/shoe size ranges from wholesale titles/descriptions.

Examples:
  S-2XL   -> ["S", "M", "L", "XL", "2XL"]
  M-3XL   -> ["M", "L", "XL", "2XL", "3XL"]
  39-45   -> ["39", "40", ..., "45"]
  sz38-45 -> ["38", "39", ..., "45"]
"""
from __future__ import annotations

import re
from typing import List

ALPHA_ORDER = ["XXS", "XS", "S", "M", "L", "XL", "2XL", "3XL", "4XL", "5XL", "6XL"]

_ALPHA_RANGE = re.compile(
    r"^(XXS|XS|S|M|L|XL|XXL|XXXL|XXXXL|\dXL)\s*[-–—]\s*"
    r"(XXS|XS|S|M|L|XL|XXL|XXXL|XXXXL|\dXL)$",
    re.I,
)
_NUM_RANGE = re.compile(r"^(\d{2})\s*[-–—]\s*(\d{2})$")
_SIZE_LABEL = re.compile(r"Sizes?\s*[:：]\s*([^.\n\r]+)", re.I)
_SZ_TOKEN = re.compile(r"\bsz\s*(\d{2}\s*[-–—]\s*\d{2})\b", re.I)


def _norm_alpha(token: str) -> str:
    t = (token or "").strip().upper().replace(" ", "")
    if not t:
        return ""
    if re.fullmatch(r"X{2,6}L", t):
        return f"{len(t) - 1}XL"
    if re.fullmatch(r"\dXL", t):
        return t
    if t in {"XXS", "XS", "S", "M", "L", "XL"}:
        return t
    return ""


def expand_size_token(raw: str) -> List[str]:
    """Expand a size token/range into discrete selectable values."""
    if not raw or not isinstance(raw, str):
        return []
    text = raw.strip().rstrip(".")
    if not text:
        return []

    m_sz = re.match(r"^sz\s*(\d{2}\s*[-–—]\s*\d{2})$", text, re.I)
    if m_sz:
        text = m_sz.group(1)

    if re.search(r"[,/|、，]", text):
        parts = [p.strip() for p in re.split(r"\s*[,/|、，]\s*", text) if p.strip()]
        if len(parts) >= 2:
            seen, out = set(), []
            for p in parts:
                if p not in seen:
                    seen.add(p)
                    out.append(p)
            return out

    m = _ALPHA_RANGE.match(text)
    if m:
        a, b = _norm_alpha(m.group(1)), _norm_alpha(m.group(2))
        if a in ALPHA_ORDER and b in ALPHA_ORDER:
            i, j = ALPHA_ORDER.index(a), ALPHA_ORDER.index(b)
            if j >= i:
                return ALPHA_ORDER[i : j + 1]

    m = _NUM_RANGE.match(text)
    if m:
        lo, hi = int(m.group(1)), int(m.group(2))
        if 20 <= lo <= hi <= 60 and hi - lo <= 20:
            return [str(n) for n in range(lo, hi + 1)]

    return []


def size_variants_from_text(*texts: str) -> List[dict]:
    """Build a structured Size variants list from free text (description/title)."""
    for text in texts:
        if not text:
            continue
        m = _SIZE_LABEL.search(text)
        token = m.group(1).strip() if m else ""
        if not token:
            m_sz = _SZ_TOKEN.search(text)
            if m_sz:
                token = m_sz.group(1)
        if not token:
            m2 = re.search(
                r"\b((?:XXS|XS|S|M|L|XL|XXL|\dXL)\s*[-–—]\s*(?:XXS|XS|S|M|L|XL|XXL|\dXL)"
                r"|\d{2}\s*[-–—]\s*\d{2})\b",
                text,
                re.I,
            )
            token = m2.group(1) if m2 else ""
        values = expand_size_token(token)
        if len(values) >= 2:
            return [{"name": "Size", "values": values}]
    return []

File: frontend/src/lib/sizeRange.js
/**
 * Expand clothing/shoe size ranges used by wholesale imports into selectable
 * values, e.g. "S-2XL" -> ["S","M","L","XL","2XL"], "39-45" -> ["39",..."45"].
 */

const ALPHA_ORDER = ['XXS', 'XS', 'S', 'M', 'L', 'XL', '2XL', '3XL', '4XL', '5XL', '6XL'];

function normAlpha(token) {
  const t = String(token || '').trim().toUpperCase().replace(/\s+/g, '');
  if (!t) return '';
  // XXL / XXXL → 2XL / 3XL
  const x = t.match(/^X{2,6}L$/);
  if (x) return `${t.length - 1}XL`;
  if (/^\dXL$/.test(t)) return t;
  if (['XXS', 'XS', 'S', 'M', 'L', 'XL'].includes(t)) return t;
  return '';
}

/**
 * Expand a single size token or range. Returns [] when not expandable.
 */
export function expandSizeToken(raw) {
  if (!raw || typeof raw !== 'string') return [];
  let text = raw.trim().replace(/\.$/, '');
  if (!text) return [];

  // Wholesale shoe/pants style: "sz38-45" / "sz 35-40"
  const sz = text.match(/^sz\s*(\d{2}\s*[-–—]\s*\d{2})$/i);
  if (sz) text = sz[1];

  // Already a list: "S, M, L" or "S/M/L"
  if (/[,/|、，]/.test(text)) {
    const parts = text
      .split(/\s*[,/|、，]\s*/)
      .map((p) => p.trim())
      .filter(Boolean);
    if (parts.length >= 2) return [...new Set(parts)];
  }

  // Alpha range: S-2XL, XS-XL, M-3XL
  const alpha = text.match(/^(XXS|XS|S|M|L|XL|XXL|XXXL|XXXXL|\dXL)\s*[-–—]\s*(XXS|XS|S|M|L|XL|XXL|XXXL|XXXXL|\dXL)$/i);
  if (alpha) {
    const a = normAlpha(alpha[1]);
    const b = normAlpha(alpha[2]);
    const i = ALPHA_ORDER.indexOf(a);
    const j = ALPHA_ORDER.indexOf(b);
    if (i >= 0 && j >= i) return ALPHA_ORDER.slice(i, j + 1);
  }

  // Numeric range: 39-45, 35-44 (shoe / pants)
  const num = text.match(/^(\d{2})\s*[-–—]\s*(\d{2})$/);
  if (num) {
    const lo = Number(num[1]);
    const hi = Number(num[2]);
    if (lo >= 20 && hi <= 60 && hi - lo <= 20 && hi >= lo) {
      const out = [];
      for (let n = lo; n <= hi; n += 1) out.push(String(n));
      return out;
    }
  }

  return [];
}

/**
 * If a Size group has a single range-like value, expand it in place.
 */
export function expandSizeGroup(group) {
  if (!group || !Array.isArray(group.values)) return group;
  if (!/^sizes?$/i.test(group.name || '')) return group;
  if (group.values.length !== 1) return group;
  const expanded = expandSizeToken(group.values[0]);
  if (expanded.length < 2) return group;
  return { ...group, values: expanded };
}
